lib/Index/Index.cpp

```diff
@@ -1048,7 +1048,7 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
       D = cast<VarDecl>(D)->getCanonicalVarDecl();
     }
 
-    if (D->isImplicit() && !isa<ConstructorDecl>(D))
+    if (!IsRef && (D->isImplicit() && !isa<ConstructorDecl>(D)))
```

Comment:
I believe the implicit check here is more for decls that are really more compiler-internal rather than the more "synthesized" decls like `hashValue`, etc. The `!isa<ConstructorDecl>` is really a hack to also include the synthesized inits. We could probably change this to `D->isImplicit() && !D->isSynthesized()` and just make sure that anything we want included is also marked synthesized - `hash(into:)` isn't right now, but `hashValue` is 😅.

The other part to that will be making sure we have a location, for decls synthesized from a derived protocol conformance IMO the best location to have there would be the conformance itself. That could be used for eg. cursor info as well, so that jumping to definition would jump to the conformance.

-----

Note that `isSynthesized` was only added a couple years ago. Its comment is:
```
    /// Whether this member was synthesized as part of a derived
    /// protocol conformance.
```

But really is more "implicit and defined for use by the user", so seems reasonable to use here.

Reply:
That sounds like a good plan. I noticed that declarations from Objective-C protocol conformances are also not marked synthesized. Do you have any tips for when/how I should set the synthesized bit? Thanks!

Reply:
My argument would be that every implicit decl that's usable by a user should be marked `synthesized`. Pinging @DougGregor in case he disagrees with that statement.

That's mostly a matter of calling `setSynthesized` wherever the node ends up being created, though perhaps `setImplicit` should take a `synthesized` arg rather than there being a separate `setSynthesized` - all synthesized are implicit, but not all implicit are synthesized.

I'm not sure where this *particular* case is, easiest way to find that is probably just to stick a conditional breakpoint (checking its name ) in the corresponding node's constructor.

Reply:
It took me a while, but I think I've updated this as suggested.

- The index store includes all decls with `synthesized` set.
- I added `synthesized` to a few more types of decls.

Most importantly, I set `synthesized` on every decl created from an Objective-C protocol requirement. Secondly, I added it to various generated decls for conformances.

I think the `_derived_(struct|enum)_equals...` doesn't exactly fit the definition of a "decl that's usable by a user", so I'm open to changing that. I'd like to follow up with some indexer changes that emit those decls and a relationship to the Decl in the `@_implements` attribute.